File: scripts/ios-native/SpotifySDKPlugin.swift
// SpotifySDKPlugin.swift
// Capacitor native plugin for iOS — mirrors android/app/src/main/.../SpotifySDKPlugin.kt
//
// Install location (after `npx cap add ios`):
//   ios/App/App/SpotifySDKPlugin.swift
//
// Requires SpotifyiOS.framework linked in the Xcode project (see docs/ios-spotify.md).
// Method return shapes MUST match the TypeScript interface in lib/spotify/capacitor-bridge.ts:
//   connect        → { connected: boolean, spotifyVersion?: string }
//   play           → { playing: boolean, trackId: string }
//   pause          → { paused: boolean }
//   resume         → { playing: boolean }
//   skipNext       → { skipped: boolean }
//   getCurrentTrack → { trackId?: string, isPlaying: boolean, position?: number }

import Foundation
import Capacitor
import SpotifyiOS

@objc(SpotifySDKPlugin)
public class SpotifySDKPlugin: CAPPlugin, SPTAppRemoteDelegate, SPTAppRemotePlayerStateDelegate {

    private var appRemote: SPTAppRemote?
    // ponytail: stored reference keeps the call alive through async Spotify auth redirect
    private var pendingConnectCall: CAPPluginCall?

    // Read from Info.plist key SPOTIFY_CLIENT_ID (set via Xcode Build Settings → User-Defined)
    private lazy var clientID: String = {
        Bundle.main.infoDictionary?["SPOTIFY_CLIENT_ID"] as? String ?? ""
    }()

    private let redirectURI = URL(string: "bingo-musical://callback")!

    override public func load() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleSpotifyCallback(_:)),
            name: .spotifyCallback,
            object: nil
        )
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleAppBecomeActive),
            name: UIApplication.didBecomeActiveNotification,
            object: nil
        )
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleAppResignActive),
            name: UIApplication.willResignActiveNotification,
            object: nil
        )
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    // Reconnect when app returns to foreground (Spotify may have disconnected)
    @objc private func handleAppBecomeActive() {
        guard let remote = appRemote, !remote.isConnected else { return }
        remote.connect()
    }

    // Disconnect when app goes to background — required by Spotify App Remote SDK
    @objc private func handleAppResignActive() {
        guard let remote = appRemote, remote.isConnected else { return }
        remote.disconnect()
    }

    // Called by AppDelegate after Spotify redirects back via bingo-musical://callback
    @objc private func handleSpotifyCallback(_ notification: Notification) {
        guard let url = notification.object as? URL,
              let params = appRemote?.authorizationParameters(from: url) else { return }

        if let token = params[SPTAppRemoteAccessTokenKey] as? String {
            appRemote?.connectionParameters.accessToken = token
            appRemote?.connect()
        } else if let errorDesc = params[SPTAppRemoteErrorDescriptionKey] as? String {
            NSLog("SpotifySDKPlugin: auth error from Spotify: %@", errorDesc)
            pendingConnectCall?.reject("Spotify auth failed: \(errorDesc)", "CONNECTION_FAILED")
            pendingConnectCall = nil
        }
    }

    // MARK: - Plugin Methods

    @objc func connect(_ call: CAPPluginCall) {
        guard !clientID.isEmpty else {
            call.reject("SPOTIFY_CLIENT_ID not set in Info.plist", "CONNECTION_FAILED")
            return
        }

        // Keep call alive until Spotify redirects back and appRemoteDidEstablishConnection fires
        call.keepAlive = true
        pendingConnectCall = call

        let config = SPTConfiguration(clientID: clientID, redirectURL: redirectURI)
        let remote = SPTAppRemote(configuration: config, logLevel: .debug)
        remote.delegate = self
        appRemote = remote

        // Must run on main thread — opens Spotify app for auth then returns via URL scheme
        DispatchQueue.main.async { remote.authorizeAndPlayURI("") }
    }

    @objc func play(_ call: CAPPluginCall) {
        guard let remote = appRemote, remote.isConnected else {
            call.reject("Spotify not connected", "NOT_CONNECTED")
            return
        }
        guard let trackId = call.getString("trackId"), !trackId.isEmpty else {
            call.reject("trackId is required", "INVALID_ARGUMENT")
            return
        }

        let contextUri = call.getString("contextUri")
        let uri = (contextUri?.isEmpty == false) ? contextUri! : "spotify:track:\(trackId)"

        remote.playerAPI?.play(uri) { _, error in
            if let error = error {
                NSLog("SpotifySDKPlugin: play error: %@", error.localizedDescription)
                call.reject("Playback failed", "PLAYBACK_FAILED")
                return
            }
            call.resolve(["playing": true, "trackId": trackId])
        }
    }

    @objc func pause(_ call: CAPPluginCall) {
        guard let remote = appRemote, remote.isConnected else {
            call.reject("Spotify not connected", "NOT_CONNECTED")
            return
        }

        remote.playerAPI?.pause { _, error in
            if let error = error {
                NSLog("SpotifySDKPlugin: pause error: %@", error.localizedDescription)
                call.reject("Pause failed", "PLAYBACK_FAILED")
                return
            }
            call.resolve(["paused": true])
        }
    }

    @objc func resume(_ call: CAPPluginCall) {
        guard let remote = appRemote, remote.isConnected else {
            call.reject("Spotify not connected", "NOT_CONNECTED")
            return
        }

        remote.playerAPI?.resume { _, error in
            if let error = error {
                NSLog("SpotifySDKPlugin: resume error: %@", error.localizedDescription)
                call.reject("Resume failed", "PLAYBACK_FAILED")
                return
            }
            call.resolve(["playing": true])
        }
    }

    @objc func skipNext(_ call: CAPPluginCall) {
        guard let remote = appRemote, remote.isConnected else {
            call.reject("Spotify not connected", "NOT_CONNECTED")
            return
        }

        remote.playerAPI?.skip(toNext: { _, error in
            if let error = error {
                NSLog("SpotifySDKPlugin: skipNext error: %@", error.localizedDescription)
                call.reject("Skip failed", "PLAYBACK_FAILED")
                return
            }
            call.resolve(["skipped": true])
        })
    }

    @objc func getCurrentTrack(_ call: CAPPluginCall) {
        guard let remote = appRemote, remote.isConnected else {
            call.reject("Spotify not connected", "NOT_CONNECTED")
            return
        }

        remote.playerAPI?.getPlayerState { result, error in
            if let error = error {
                NSLog("SpotifySDKPlugin: getCurrentTrack error: %@", error.localizedDescription)
                call.reject("Failed to get player state", "PLAYBACK_FAILED")
                return
            }

            guard let state = result as? SPTAppRemotePlayerState else {
                call.resolve(["isPlaying": false])
                return
            }

            var res: JSObject = ["isPlaying": !state.isPaused]
            if let track = state.track {
                res["trackId"] = track.uri.replacingOccurrences(of: "spotify:track:", with: "")
                res["position"] = state.playbackPosition
            }
            call.resolve(res)
        }
    }

    // MARK: - SPTAppRemoteDelegate

    public func appRemoteDidEstablishConnection(_ appRemote: SPTAppRemote) {
        appRemote.playerAPI?.delegate = self
        pendingConnectCall?.resolve(["connected": true])
        pendingConnectCall = nil
    }

    public func appRemote(_ appRemote: SPTAppRemote, didDisconnectWithError error: Error?) {
        if let error = error {
            NSLog("SpotifySDKPlugin: disconnected: %@", error.localizedDescription)
        }
    }

    public func appRemote(_ appRemote: SPTAppRemote, didFailConnectionAttemptWithError error: Error?) {
        let msg = error?.localizedDescription ?? "Unknown error"
        NSLog("SpotifySDKPlugin: connection failed: %@", msg)

        let msgLower = msg.lowercased()
        if msgLower.contains("not installed") || msgLower.contains("not found") || msgLower.contains("spotify") {
            pendingConnectCall?.reject("Spotify app not installed", "APP_NOT_INSTALLED")
        } else {
            pendingConnectCall?.reject("Failed to connect to Spotify", "CONNECTION_FAILED")
        }
        pendingConnectCall = nil
    }

    // MARK: - SPTAppRemotePlayerStateDelegate

    public func playerStateDidChange(_ playerState: SPTAppRemotePlayerState) {
        // State changes are handled per-call; no persistent subscription needed
    }
}

// MARK: - Notification name shared with AppDelegate.additions.swift
extension Notification.Name {
    static let spotifyCallback = Notification.Name("SpotifyCallbackURL")
}

File: android/app/src/main/java/com/bingomusical/app/SpotifySDKPlugin.kt
package com.bingomusical.app

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.spotify.android.appremote.api.ConnectionParams
import com.spotify.android.appremote.api.Connector
import com.spotify.android.appremote.api.SpotifyAppRemote
import com.spotify.protocol.types.Track

/**
 * SpotifySDKPlugin — Capacitor native plugin for Android
 *
 * Implements the CapacitorSpotifyPlugin interface defined in lib/spotify/capacitor-bridge.ts.
 * Controls Spotify playback via the Spotify App Remote SDK (com.spotify.android:app-remote:0.8.0).
 *
 * Method return shapes MUST match the TypeScript interface exactly:
 *   connect  → { connected: boolean, spotifyVersion?: string }
 *   play     → { playing: boolean, trackId: string }
 *   pause    → { paused: boolean }
 *   resume   → { playing: boolean }
 *   skipNext → { skipped: boolean }
 *   getCurrentTrack → { trackId?: string, isPlaying: boolean, position?: number }
 */
@CapacitorPlugin(name = "SpotifySDK")
class SpotifySDKPlugin : Plugin() {

    private var spotifyAppRemote: SpotifyAppRemote? = null

    // SPOTIFY_CLIENT_ID is injected at Gradle build time from NEXT_PUBLIC_SPOTIFY_CLIENT_ID env var.
    // It is NOT hardcoded in source — see android/app/build.gradle buildConfigField.
    private val CLIENT_ID: String = BuildConfig.SPOTIFY_CLIENT_ID
    private val REDIRECT_URI = "bingo-musical://callback"

    /**
     * Connect to the Spotify app via App Remote SDK.
     *
     * On emulators without Spotify installed, onFailure is called immediately with
     * a message containing "CouldNotFindSpotifyApp". The bridge returns APP_NOT_INSTALLED
     * so the JS layer can set shouldFallback: true.
     *
     * Returns: { connected: true } on success
     * Rejects: "APP_NOT_INSTALLED" | "CONNECTION_FAILED"
     */
    @PluginMethod
    fun connect(call: PluginCall) {
        val connectionParams = ConnectionParams.Builder(CLIENT_ID)
            .setRedirectUri(REDIRECT_URI)
            .showAuthView(true)
            .build()

        SpotifyAppRemote.connect(
            context,
            connectionParams,
            object : Connector.ConnectionListener {
                override fun onConnected(appRemote: SpotifyAppRemote) {
                    spotifyAppRemote = appRemote
                    val ret = JSObject()
                    ret.put("connected", true)
                    call.resolve(ret)
                }

                override fun onFailure(throwable: Throwable) {
                    val message = throwable.message ?: "Unknown error"
                    // Log details server-side only — do not expose internals to client
                    bridge.activity.runOnUiThread {
                        android.util.Log.e("SpotifySDKPlugin", "connect onFailure: $message", throwable)
                    }
                    if (message.contains("AUTHENTICATION_SERVICE_UNAVAILABLE", ignoreCase = true) ||
                        message.contains("CouldNotFindSpotifyApp", ignoreCase = true) ||
                        message.contains("SpotifyDisconnectedException", ignoreCase = true)
                    ) {
                        call.reject("APP_NOT_INSTALLED", "Spotify app not installed")
                    } else {
                        call.reject("CONNECTION_FAILED", "Failed to connect to Spotify")
                    }
                }
            }
        )
    }

    /**
     * Play a track by trackId or contextUri.
     *
     * Input: { trackId: string, contextUri?: string }
     * Returns: { playing: true, trackId: string }
     * Rejects: "NOT_CONNECTED" | "INVALID_ARGUMENT" | "PLAYBACK_FAILED"
     */
    @PluginMethod
    fun play(call: PluginCall) {
        val remote = spotifyAppRemote
        if (remote == null || !remote.isConnected) {
            call.reject("NOT_CONNECTED", "Spotify not connected")
            return
        }

        val trackId = call.getString("trackId")
        if (trackId.isNullOrBlank()) {
            call.reject("INVALID_ARGUMENT", "trackId is required")
            return
        }

        val contextUri = call.getString("contextUri")
        val spotifyUri = if (!contextUri.isNullOrBlank()) contextUri else "spotify:track:$trackId"

        remote.playerApi.play(spotifyUri)
            .setResultCallback {
                val ret = JSObject()
                ret.put("playing", true)
                ret.put("trackId", trackId)
                call.resolve(ret)
            }
            .setErrorCallback { throwable ->
                android.util.Log.e("SpotifySDKPlugin", "play error: ${throwable.message}", throwable)
                call.reject("PLAYBACK_FAILED", "Playback failed")
            }
    }

    /**
     * Pause current playback.
     *
     * Returns: { paused: true }
     * Rejects: "NOT_CONNECTED" | "PLAYBACK_FAILED"
     */
    @PluginMethod
    fun pause(call: PluginCall) {
        val remote = spotifyAppRemote
        if (remote == null || !remote.isConnected) {
            call.reject("NOT_CONNECTED", "Spotify not connected")
            return
        }

        remote.playerApi.pause()
            .setResultCallback {
                val ret = JSObject()
                ret.put("paused", true)
                call.resolve(ret)
            }
            .setErrorCallback { throwable ->
                android.util.Log.e("SpotifySDKPlugin", "pause error: ${throwable.message}", throwable)
                call.reject("PLAYBACK_FAILED", "Pause failed")
            }
    }

    /**
     * Resume paused playback.
     *
     * Returns: { playing: true }
     * Rejects: "NOT_CONNECTED" | "PLAYBACK_FAILED"
     */
    @PluginMethod
    fun resume(call: PluginCall) {
        val remote = spotifyAppRemote
        if (remote == null || !remote.isConnected) {
            call.reject("NOT_CONNECTED", "Spotify not connected")
            return
        }

        remote.playerApi.resume()
            .setResultCallback {
                val ret = JSObject()
                ret.put("playing", true)
                call.resolve(ret)
            }
            .setErrorCallback { throwable ->
                android.util.Log.e("SpotifySDKPlugin", "resume error: ${throwable.message}", throwable)
                call.reject("PLAYBACK_FAILED", "Resume failed")
            }
    }

    /**
     * Skip to next track.
     *
     * Returns: { skipped: true }
     * Rejects: "NOT_CONNECTED" | "PLAYBACK_FAILED"
     */
    @PluginMethod
    fun skipNext(call: PluginCall) {
        val remote = spotifyAppRemote
        if (remote == null || !remote.isConnected) {
            call.reject("NOT_CONNECTED", "Spotify not connected")
            return
        }

        remote.playerApi.skipNext()
            .setResultCallback {
                val ret = JSObject()
                ret.put("skipped", true)
                call.resolve(ret)
            }
            .setErrorCallback { throwable ->
                android.util.Log.e("SpotifySDKPlugin", "skipNext error: ${throwable.message}", throwable)
                call.reject("PLAYBACK_FAILED", "Skip failed")
            }
    }

    /**
     * Get info about the currently playing track.
     *
     * Returns: { isPlaying: boolean, trackId?: string, position?: number }
     * Rejects: "NOT_CONNECTED" | "PLAYBACK_FAILED"
     */
    @PluginMethod
    fun getCurrentTrack(call: PluginCall) {
        val remote = spotifyAppRemote
        if (remote == null || !remote.isConnected) {
            call.reject("NOT_CONNECTED", "Spotify not connected")
            return
        }

        remote.playerApi.playerState
            .setResultCallback { state ->
                val track: Track? = state.track
                val ret = JSObject()
                ret.put("isPlaying", !state.isPaused)
                if (track != null) {
                    // Strip "spotify:track:" prefix to return only the track ID
                    ret.put("trackId", track.uri.removePrefix("spotify:track:"))
                    ret.put("position", state.playbackPosition)
                }
                call.resolve(ret)
            }
            .setErrorCallback { throwable ->
                android.util.Log.e("SpotifySDKPlugin", "getCurrentTrack error: ${throwable.message}", throwable)
                call.reject("PLAYBACK_FAILED", "Failed to get player state")
            }
    }

    /**
     * Disconnect from Spotify App Remote when the plugin is destroyed.
     * Called by Capacitor when the Activity is destroyed.
     */
    override fun handleOnDestroy() {
        SpotifyAppRemote.disconnect(spotifyAppRemote)
        spotifyAppRemote = null
        super.handleOnDestroy()
    }
}
